cleanup_utils/openstack_cleanup.py

```diff
@@ -0,0 +1,135 @@
+
+
+# Classes #####################################################################
+
+class OpenStackCleanupInstance:
+    """
+    Handles the search and cleanup of all unused OpenStack resources used
+    by CircleCI
+    """
+    def __init__(self, age_limit, openstack_settings, dry_run=False):
+        """
+        Set's up Nova client
+        """
+        self.dry_run = dry_run
+        self.age_limit = age_limit
+        self.cleaned_up_hashes = []
+
+        self.nova = client.Client(
+            "2.0",
+            auth_url=openstack_settings['auth_url'],
+            username=openstack_settings['username'],
+            api_key=openstack_settings['api_key'],
+            project_id=openstack_settings['project_id'],
+            region_name=openstack_settings['region_name']
+        )
+
+    def get_active_circle_ci_instances(self):
+        """
+        Returns list of active instances running on the OpenStack provider
+        that have been created by CircleCI
+
+        Note: key_name parameter is useless if ran with a user that is not an
+              admin. It's good to check twice somewhere else.
+        """
+        server_list = self.nova.servers.list(
+            search_opts={
+                'key_name': 'circleci'
+            }
+        )
+        # Filter any keys that are not from circleci in case our query
+        # was ignored
+        server_list = [s for s in server_list if s.key_name == 'circleci']
+        return server_list
+
+    def run_cleanup(self):
+        """
+        Runs the cleanup of OpenStack provider
+        """
+        logger.info("\n --- Starting OpenStack Provider Cleanup ---")
+        if self.dry_run:
+            logger.info("Running in DRY_RUN mode, no actions will be taken.")
+
+        ci_instances = self.get_active_circle_ci_instances()
+        logger.info("Found %s active instances using CircleCI keys...", len(ci_instances))
+
+        for instance in ci_instances:
+            logger.info("  > Checking instance %s...", instance.name)
+            logger.info(
+                "    * id=%s, key_name=%s, created=%s",
+                instance.id,
+                instance.key_name,
+                instance.created
+            )
+            # Double-check to make sure that the instance is using the circleci keypair.
+            if instance.key_name != 'circleci':
+                logger.info(
+                    "    * SKIPPING: Instance keypair name %s != 'circleci'!",
+                    instance.key_name
+                )
+                continue
+
+            # Check if it's a valid date and add UTC timezone
+            try:
+                instance_age_unaware = datetime.strptime(
+                    instance.created,
+                    '%Y-%m-%dT%H:%M:%SZ'
+                )
+                instance_age = instance_age_unaware.replace(tzinfo=pytz.UTC)
+            except ValueError:
+                instance_age = None
```

Comment:
@giovannicimolin, it might be better to log a warning as the logging code below in the else block will happily convert `None` to `'None'` and it might be difficult to trace the log in case of issues.

Reply:
@giovannicimolin, it will get to the `else` block because `instance_age = None` is falsey.